File: Castro-Lara-Daniel_Ivan/Procesos/files C++/pipes.cpp
/*
    Sistemas Operativos
    Daniel Iván Castro Lara
    Programa en C++
	$ gcc -o procesos_v0_param procesos_v0_param.cpp
*/

#include <iostream>
#include <unistd.h>   // Para pipe(), write()
#include <cstring>    // Para strlen()
#include <cerrno>     // Para errno
#include <cstdio>     // Para perror()

int main() {
    int fd[2];                  // Descriptores del pipe: fd[0] para lectura, fd[1] para escritura
    const char *buf = "a";      // Buffer que se escribirá en el pipe

    // Crear el pipe
    if (pipe(fd) < 0) {
        std::cerr << "Error en la creación del pipe: " << strerror(errno) << std::endl;
        return 1;
    }

    for (int i = 1; ; i++) {
        if (write(fd[1], buf, 1) < 0) {
            std::cerr << "Error al escribir en el pipe: " << strerror(errno) << std::endl;
            break;
        }

        std::cout << i << " ";
        std::cout.flush();  // Vaciar el búfer de salida para mostrar los números inmediatamente
    }

    // Cerrar los descriptores del pipe
    close(fd[0]);
    close(fd[1]);

    return 0;
}

File: Castro-Lara-Daniel_Ivan/Procesos/files Python/pipes.py
"""
    Sistemas Operativos
    Daniel Iván Castro Lara
    Programa en Python 
"""

import os

def main():
    # Crear un pipe
    fd_read, fd_write = os.pipe()
    buf = b"a"  # El buffer debe ser de tipo bytes

    try:
        i = 1
        while True:
            os.write(fd_write, buf)  # Escribir un byte en el extremo de escritura del pipe
            print(i, end=' ', flush=True)
            i += 1
    except KeyboardInterrupt:
        print("\nEjecución interrumpida por el usuario")
    finally:
        # Cerrar los extremos del pipe
        os.close(fd_read)
        os.close(fd_write)

if __name__ == "__main__":
    main()

"""
    1.- python <archivo>.py
    No se necesita compilar porque es interpretado.
"""
